src/qiskit_qasm3_import/state.py

```diff
@@ -0,0 +1,233 @@
+
+    def __init__(self):
+        self._state = self.UNKNOWN
+
+    def set_physical_mode(self, node):
+        """Set the addressing mode to physical. On success return `True`, otherwise raise an exception."""
+        if self._state == self.VIRTUAL:
+            raise_from_node(
+                node,
+                "Physical qubit referenced in virtual addressing mode. Mixing modes not currently supported.",
+            )
+        self._state = self.PHYSICAL
+
+    def set_virtual_mode(self, node):
+        """Set the addressing mode to virtual. On success return `True`, otherwise raise an exception."""
+        if self._state == self.PHYSICAL:
+            raise_from_node(
+                node,
+                "Virtual qubit declared in physical addressing mode. Mixing modes not currently supported.",
+            )
+        self._state = self.VIRTUAL
+
+
+class SymbolTable:
+    __slots__ = ("_local_table", "_global_table", "_builtin_table")
+
+    def __init__(self):
+        self._global_table = {}
+        self._builtin_table = _BUILTINS.copy()
+        self._local_table = {}  # For everything else
+
+    def gate_scope_copy(self):
+        """Return a copy of the symbol table for use in the lexical scope of a gate definition.
+
+        The target (returned) symbol table contains: all builtin symbols, no symbols from a surrounding local scope,
+        all symbols from the global scope referring to gates gates, all symbols from the global scope
+        that are marked constant and refer to numeric data.
+        """
+        # pylint: disable=protected-access
+        out = SymbolTable.__new__(SymbolTable)
+        out._local_table = {}
+        out._builtin_table = self._builtin_table
+        out._global_table = {}
+        for name, item in self._global_table.items():
+            if (
+                isinstance(item.type, types.Gate)
+                or isinstance(
+                    item.type, (types.Int, types.Uint, types.Float, types.Angle, types.Duration)
+                )
+                and item.type.const
+            ):
+                out._global_table[name] = item  # TODO: use insert
+        return out
+
+    def local_scope_copy(self):
+        """Return a copy of the symbol table for use with a new local scope.
+
+
+        Local variables created in the target (returned) symbol table will not
+        appear in the source symbol table. Thus, these local variables are
+        discarded and not visible upon returning to the surrounding scope. In
+        contrast, changes to data referred to by global symbols will be visible
+        in the surrounding scope.
+        """
+        # pylint: disable=protected-access
+        out = SymbolTable.__new__(SymbolTable)
+        out._local_table = self._local_table.copy()
+        out._global_table = self._global_table
+        out._builtin_table = self._builtin_table
+
+        return out
```

Comment:
I agree this design is more appealing. This has a more stack-like structure. And so do scopes. And it's probably more efficient because it avoids copying. It might be more efficient when iterating over symbols to separate some kinds of symbols, e.g. hardware qubits. But thats an optimization.

Also, you inadvertently omitted looking for symbols in `global_symbols`, right?

Reply:
I don't think I omitted any lookup here?  The table first tries to look in itself (the local scope).  If that fails and there's no parent scope, it complains "symbol not found".  Otherwise it asks its parent scope to try a lookup.  Once it's got a match it finally checks to see if it's allowed to "see" that symbol.  The recursive `self.base` lookup should handle going all the way up to the global scope?

I don't think it's a good idea to try and separate out various symbols without a good cause. It opens more surface area for bugs (you need to check more places for definitions / keep more data in sync), and the number of symbols we're going to have to search through really isn't going to be all that high.

Reply:
In any case, at least for the first iteration, I think it's best to have a simple implementation. It doesn't look like it will be wasted effort if we need to make it more complicated later.